pandas/core/sparse/array.py

```diff
@@ -165,253 +206,282 @@ def __new__(cls, data, sparse_index=None, index=None, kind='integer',
+                    raise
 
-        if is_sparse_array:
-            sparse_index = data.sp_index
-            values = data.sp_values
-            fill_value = data.fill_value
-        else:
-            # array-like
-            if sparse_index is None:
-                if dtype is not None:
-                    data = np.asarray(data, dtype=dtype)
-                res = make_sparse(data, kind=kind, fill_value=fill_value)
-                values, sparse_index, fill_value = res
-            else:
-                values = _sanitize_values(data)
-                if len(values) != sparse_index.npoints:
-                    raise AssertionError("Non array-like type {type} must "
-                                         "have the same length as the index"
-                                         .format(type=type(values)))
-        # Create array, do *not* copy data by default
         if copy:
-            subarr = np.array(values, dtype=dtype, copy=True)
-        else:
-            subarr = np.asarray(values, dtype=dtype)
-        # Change the class of the array to be the subclass type.
-        return cls._simple_new(subarr, sparse_index, fill_value)
-
-    @classmethod
-    def _simple_new(cls, data, sp_index, fill_value):
-        if not isinstance(sp_index, SparseIndex):
-            # caller must pass SparseIndex
-            raise ValueError('sp_index must be a SparseIndex')
+            # TODO: avoid double copy when dtype forces cast.
+            data = data.copy()
 
         if fill_value is None:
-            if sp_index.ngaps > 0:
-                # has missing hole
+            fill_value_dtype = dtype or data.dtype
+            if fill_value_dtype is None:
                 fill_value = np.nan
             else:
-                fill_value = na_value_for_dtype(data.dtype)
+                fill_value = na_value_for_dtype(fill_value_dtype)
+
+        if isinstance(data, type(self)) and sparse_index is None:
+            sparse_index = data._sparse_index
+            sparse_values = np.asarray(data.sp_values, dtype=dtype)
+        elif sparse_index is None:
+            sparse_values, sparse_index, fill_value = make_sparse(
+                data, kind=kind, fill_value=fill_value, dtype=dtype
+            )
+        else:
+            sparse_values = np.asarray(data, dtype=dtype)
+            if len(sparse_values) != sparse_index.npoints:
+                raise AssertionError("Non array-like type {type} must "
+                                     "have the same length as the index"
+                                     .format(type=type(sparse_values)))
+        # TODO: copy is unused
+
+        self._sparse_index = sparse_index
+        self._sparse_values = sparse_values
+        self._dtype = SparseDtype(sparse_values.dtype)
+        self.fill_value = fill_value
+
+    @classmethod
+    def _simple_new(cls, sparse_array, sparse_index, fill_value=None):
+        # type: (SparseArray, SparseIndex) -> 'SparseArray'
```

Comment:
SparseArray -> ndarray ?

Reply:
I messed something up here... Will look how it's used.

It should probably be ndarray though, and I'll have to adjust the fill_value code (which is clearly untested).